test/unittest/dds/subscriber/DataReaderTests.cpp

```diff
@@ -12,152 +12,787 @@
+    {
+        EXPECT_EQ(wrong_code, data_reader_->read(test.first, test.second));
+        EXPECT_EQ(wrong_code, data_reader_->read_next_instance(test.first, test.second));
+        EXPECT_EQ(wrong_code, data_reader_->take(test.first, test.second));
+        EXPECT_EQ(wrong_code, data_reader_->take_next_instance(test.first, test.second));
+
+        check_instance_methods(wrong_code, wrong_code, wrong_code,
+                data_reader_, test.first, test.second);
     }
 
-    bool getKey(
-            void* /*data*/,
-            fastrtps::rtps::InstanceHandle_t* /*ihandle*/,
-            bool /*force_md5*/) override
+    // Check compatible combinations
+    using ok_test_case_t = std::pair<test_case_t, const ReturnCode_t&>;
+    std::vector<ok_test_case_t> ok_cases
     {
-        return true;
+        // max == 0. Loaned data will be returned.
+        { {true_0_0, info_true_0_0}, ok_code},
+        // max > 0 && owns == true. Data will be copied.
+        { {true_10_0, info_true_10_0}, ok_code},
+        { {true_10_1, info_true_10_1}, ok_code},
+        // max > 0 && owns == false. Precondition not met.
+        { {false_10_0, info_false_10_0}, wrong_code},
+        { {false_10_1, info_false_10_1}, wrong_code}
+    };
+
+    const ReturnCode_t& instance_bad_code = ReturnCode_t::RETCODE_BAD_PARAMETER;
+    for (const ok_test_case_t& test : ok_cases)
+    {
+        EXPECT_EQ(test.second, data_reader_->read(test.first.first, test.first.second));
+        EXPECT_EQ(wrong_code, data_reader_->return_loan(test.first.first, test.first.second));
+        EXPECT_EQ(test.second, data_reader_->read_next_instance(test.first.first, test.first.second));
+        EXPECT_EQ(wrong_code, data_reader_->return_loan(test.first.first, test.first.second));
+        EXPECT_EQ(test.second, data_reader_->take(test.first.first, test.first.second));
+        EXPECT_EQ(wrong_code, data_reader_->return_loan(test.first.first, test.first.second));
+        EXPECT_EQ(test.second, data_reader_->take_next_instance(test.first.first, test.first.second));
+        EXPECT_EQ(wrong_code, data_reader_->return_loan(test.first.first, test.first.second));
+
+        // When collection preconditions are ok, as the reader has no data, BAD_PARAMETER will be returned
+        const ReturnCode_t& instance_code = (test.second == ok_code) ? instance_bad_code : test.second;
+        check_instance_methods(instance_code, instance_code, wrong_code,
+                data_reader_, test.first.first, test.first.second);
     }
 
-};
+    false_10_0.unloan();
+    false_10_1.unloan();
+    info_false_10_0.unloan();
+    info_false_10_1.unloan();
+}
 
-TEST(DataReaderTests, ReadData)
+void forward_loan(
+        LoanableCollection& to,
+        const LoanableCollection& from)
 {
-    DomainParticipant* participant =
-            DomainParticipantFactory::get_instance()->create_participant(0, PARTICIPANT_QOS_DEFAULT);
-    ASSERT_NE(participant, nullptr);
+    assert(from.has_ownership() == false);
```

Comment:
I would add a comment in the assert to make the error more traceable

Reply:
Done in 5a52b6655364ed88e018c0863b8d2f72606383a0